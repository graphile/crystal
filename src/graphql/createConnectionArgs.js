import { memoize, fromPairs, includes, snakeCase, toUpper } from 'lodash'
import { GraphQLEnumType, GraphQLInt, GraphQLBoolean } from 'graphql'
import { CursorType } from './types.js'
import getType from './getType.js'

const createConnectionArgs = (table, ignoreColumnConditions = []) => ({
  // The column specified by `orderBy` means more than just the order to
  // return items in. This column is also the column we will use for
  // cursors.
  orderBy: {
    type: createTableOrderingEnum(table),
    description:
      'The order the resulting items should be returned in. This argument ' +
      'is also important as it is used in creating pagination cursors. This ' +
      'value’s default is the primary key for the object.',
    defaultValue: (() => {
      const column = table.getPrimaryKeys()[0]
      if (column) return column.name
      return null
    })(),
  },
  first: {
    type: GraphQLInt,
    description:
      'The top `n` items in the set to be returned. Can’t be used ' +
      'with `last`.',
  },
  last: {
    type: GraphQLInt,
    description:
      'The bottom `n` items in the set to be returned. Can’t be used ' +
      'with `first`.',
  },
  before: {
    type: CursorType,
    description: 'Constrains the set to nodes *before* this cursor in the specified ordering.',
  },
  after: {
    type: CursorType,
    description: 'Constrains the set to nodes *after* this cursor in the specified ordering.',
  },
  offset: {
    type: GraphQLInt,
    description: 'An integer offset representing how many items to skip in the set.',
  },
  descending: {
    type: GraphQLBoolean,
    description:
      'If `true` the nodes will be in descending order, if `false` the ' +
      'items will be in ascending order. `false` by default.',
    defaultValue: false,
  },
  ...fromPairs(
    table
    .getColumns()
    // If `ignoreColumnConditions` is set to true, all column conditions will
    // be disabled. If `ignoreColumnConditions` is an array, only certain
    // conditions will be ignored.
    .filter(column => (ignoreColumnConditions === true ? false : !includes(ignoreColumnConditions, column)))
    .map(column => [column.getFieldName(), {
      type: getType(column.type),
      description:
        'Filters the resulting set with an equality test on the ' +
        `${column.getMarkdownFieldName()} field.`,
      // TODO: Deprecate this…
      // deprecationReason:
      //   'Simple equality testing is insufficient for the nodes field and just ' +
      //   'adds noise. Instead use procedures for custom set filtering.',
    }])
  ),
})

export default createConnectionArgs

/**
 * Creates an ordering enum which simply contains all of a `Table`s columns.
 *
 * @param {Table} table
 * @returns {GraphQLEnumType}
 */
// TODO: Some way to eliminate some columns from ordering enum?
export const createTableOrderingEnum = memoize(table =>
  new GraphQLEnumType({
    name: `${table.getTypeName()}Ordering`,
    description: `Properties with which ${table.getMarkdownTypeName()} can be ordered.`,

    values: fromPairs(
      table
      .getColumns()
      .map(column => [toUpper(snakeCase(column.getFieldName())), {
        value: column.name,
        description: column.description,
      }])
    ),
  }))
