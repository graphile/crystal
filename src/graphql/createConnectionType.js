import { memoize } from 'lodash'
import createTableType from './createTableType.js'
import { PageInfoType, CursorType } from './types.js'

import {
  GraphQLObjectType,
  GraphQLList,
  GraphQLNonNull,
  GraphQLInt,
} from 'graphql'

const createTableConnectionType = memoize(table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Connection`,
    description: `A connection to a list of ${table.getMarkdownTypeName()} items`,

    // TODO: Implement a `ConnectionType` interface

    fields: {
      pageInfo: {
        type: new GraphQLNonNull(PageInfoType),
        description: `Information to aid in pagination of type ${table.getMarkdownTypeName()}.`,
        resolve: pageInfo => pageInfo,
      },
      totalCount: {
        type: GraphQLInt,
        description: 'All of the items available to be queried in this connection.',
        resolve: ({ totalCount }) => totalCount,
      },
      nodes: {
        type: new GraphQLList(createTableType(table)),
        description: `The queried list of ${table.getMarkdownTypeName()}.`,
        resolve: ({ nodes }) => nodes,
      },
      edges: {
        type: new GraphQLList(createTableEdgeType(table)),
        description: 'A single item and a cursor to aid in pagination.',
        resolve: ({ edges }) => edges,
      },
    },
  })
)

export default createTableConnectionType

const createTableEdgeType = table =>
  new GraphQLObjectType({
    name: `${table.getTypeName()}Edge`,
    description: `An edge in the \`${table.getTypeName()}Connection\`.`,

    fields: {
      cursor: {
        type: new GraphQLNonNull(CursorType),
        description: 'The cursor describing the position of the edgeâ€™s associated node.',
        resolve: ({ cursor }) => cursor || 'null',
      },
      node: {
        type: createTableType(table),
        description: 'The item at the end of the edge.',
        resolve: ({ node }) => node,
      },
    },
  })
