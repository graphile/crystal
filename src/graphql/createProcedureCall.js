const createProcedureCall = procedure => {
  // Construct the qualified procedure name.
  const procedureName = `"${procedure.schema.name}"."${procedure.name}"`

  // Construct the argument list for the procedure call.
  const procedureArgsList = Array.from(procedure.args).map(([name, type], i) => {
    const placeholder = `$${i + 1}`

    // If the type of this argument is a table type, we will be expecting JSON
    // but we need that JSON to be a table. Therefore we run it through
    // `json_populate_record`.
    if (type.isTableType)
      return `json_populate_record(null::"${type.table.schema.name}"."${type.table.name}", ${placeholder})`

    return placeholder
  }).join(', ')

  // Add the procedure name with the procedure argument list.
  const procedureCall = `${procedureName}(${procedureArgsList})`

  return procedureCall
}

export default createProcedureCall
