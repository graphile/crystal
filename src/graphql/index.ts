import createGraphQLSchema from './schema/createGQLSchema'

export {
  createGraphQLSchema,
}
