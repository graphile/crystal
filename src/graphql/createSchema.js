import { GraphQLSchema } from 'graphql'
import createQueryType from './query/createQueryType.js'
import createMutationType from './mutation/createMutationType.js'

/**
 * Creates a GraphQLSchema from a PostgreSQL schema.
 *
 * @param {Schema} schema
 * @returns {GrpahQLSchema}
 */
const createGraphqlSchema = (schema, options) =>
  new GraphQLSchema({
    query: createQueryType(schema, options),
    mutation: createMutationType(schema, options),
  })

export default createGraphqlSchema
