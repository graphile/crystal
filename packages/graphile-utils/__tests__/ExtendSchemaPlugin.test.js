import {
  makeExtendSchemaPlugin,
  makeAddInflectorsPlugin,
  gql,
  embed,
} from "../";
import {
  buildSchema,
  // defaultPlugins,
  StandardTypesPlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  MutationPayloadQueryPlugin,
} from "graphile-build";
import { graphql, subscribe, parse } from "graphql";
import { $$asyncIterator } from "iterall";

function TestUtils_ExtractScopePlugin(
  hook,
  objectTypeName,
  fieldNameOrCallback,
  possiblyCallback,
) {
  const callback =
    typeof fieldNameOrCallback === "function" && !possiblyCallback
      ? fieldNameOrCallback
      : possiblyCallback;
  const fieldName = possiblyCallback && fieldNameOrCallback;
  return (builder) => {
    builder.hook(hook, (_, build, context) => {
      const { Self } = context;
      const currentObjectTypeName = (Self && Self.name) || _.name;
      const currentFieldName = Self ? context.scope.fieldName : undefined;
      if (
        currentObjectTypeName === objectTypeName &&
        (!fieldName || fieldName === currentFieldName)
      ) {
        callback(context.scope);
      }
      return _;
    });
  };
}

const simplePlugins = [
  StandardTypesPlugin,
  QueryPlugin,
  MutationPlugin,
  SubscriptionPlugin,
  MutationPayloadQueryPlugin,
];

let timerRunning = false;
const resolvers = {
  Query: {
    randomNumber(_query, _args, _context, _info) {
      return 4; // chosen by fair dice roll. guaranteed to be random. xkcd#221
    },
    randomNumbers() {
      return [5, 3, 6];
    },
    echo(_query, args) {
      return args.input;
    },
  },
  Mutation: {
    add(_mutation, args) {
      const { a, b } = args;
      // So this isn't a mutation. Whatever.
      return a + b;
    },
  },
  Subscription: {
    clockTicks: {
      resolve(_subscription, args) {
        const { frequency } = args;
        if (frequency == null) {
          throw new Error("No frequency specified");
        }
        return new Promise((resolve) =>
          setTimeout(() => resolve(Date.now()), frequency),
        );
      },
      subscribe(_subscription, args) {
        const { frequency } = args;
        if (frequency == null) {
          throw new Error("No frequency specified");
        }
        const callbackQueue = [];
        const valueQueue = [];
        // In a normal application you'd define timerRunning here:
        //   const timerRunning = true
        // However, in the tests we want access to this variable so I've moved
        // it to the global scope.
        timerRunning = true;
        function addValue(v) {
          if (!timerRunning) {
            return;
          }
          if (callbackQueue.length) {
            const callback = callbackQueue.shift();
            callback({ value: v, done: false });
          } else {
            valueQueue.push(v);
          }
        }
        function nextValue() {
          if (valueQueue.length) {
            return Promise.resolve(valueQueue.shift());
          } else {
            return new Promise((resolve) => {
              callbackQueue.push(resolve);
            });
          }
        }
        const interval = setInterval(() => addValue(Date.now()), frequency);
        function stopIterator() {
          if (timerRunning) {
            timerRunning = false;
            clearInterval(interval);
          }
        }
        return {
          next() {
            return timerRunning ? nextValue() : this.return();
          },
          return() {
            stopIterator();
            return Promise.resolve({ value: undefined, done: true });
          },
          throw(error) {
            stopIterator();
            return Promise.reject(error);
          },
          [$$asyncIterator]() {
            return this;
          },
        };
      },
    },
  },
};

it("allows adding a simple type", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          A random number generated by a fair dice roll.
          """
          randomNumber: Int
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        randomNumber
      }
    `,
  );
  expect(data.randomNumber).toEqual(4);
});

it("allows adding a non-null type", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          A random number generated by a fair dice roll.
          """
          randomNumber: Int!
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        randomNumber
      }
    `,
  );
  expect(data.randomNumber).toEqual(4);
});

it("allows adding a non-null list of non-null type", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          Gives a list of numbers that were randomly generated by fair dice roll
          """
          randomNumbers: [Int!]!
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        randomNumbers
      }
    `,
  );
  expect(data.randomNumbers).toEqual([5, 3, 6]);
});

it("accepts an array of typedefs", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: [
        gql`
          extend type Query {
            """
            A random number generated by a fair dice roll.
            """
            randomNumber: Int!
          }
        `,
        gql`
          extend type Query {
            """
            Gives a list of numbers that were randomly generated by fair dice roll
            """
            randomNumbers: [Int!]!
          }
        `,
      ],
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        randomNumber
        randomNumbers
      }
    `,
  );
  expect(data.randomNumber).toEqual(4);
  expect(data.randomNumbers).toEqual([5, 3, 6]);
});

it("throws the proper error if an array of typeDefs aren't all DocumentNodes", () => {
  return expect(
    buildSchema([
      ...simplePlugins,
      makeExtendSchemaPlugin((_build) => ({
        typeDefs: [
          gql`
            extend type Query {
              """
              A random number generated by a fair dice roll.
              """
              randomNumber: Int!
            }
          `,
          `
            extend type Query {
              """
              Gives a list of numbers that were randomly generated by fair dice roll
              """
              randomNumbers: [Int!]!
            }
          `,
        ],
        resolvers,
      })),
    ]),
  ).rejects.toMatchInlineSnapshot(
    `[Error: The first argument to makeExtendSchemaPlugin must be generated by the \`gql\` helper, or be an array of the same.]`,
  );
});

it("throws the proper error if a single typeDef isn't a DocumentNode", () => {
  return expect(
    buildSchema([
      ...simplePlugins,
      makeExtendSchemaPlugin((_build) => ({
        typeDefs: `
            extend type Query {
              """
              Gives a list of numbers that were randomly generated by fair dice roll
              """
              randomNumbers: [Int!]!
            }
          `,
        resolvers,
      })),
    ]),
  ).rejects.toMatchInlineSnapshot(
    `[Error: The first argument to makeExtendSchemaPlugin must be generated by the \`gql\` helper, or be an array of the same.]`,
  );
});

it("allows adding a field with arguments", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          Gives you back what you put in
          """
          echo(input: [Int!]!): [Int!]!
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data } = await graphql(
    schema,
    `
      {
        echo(input: [1, 1, 2, 3, 5, 8])
      }
    `,
  );
  expect(data.echo).toEqual([1, 1, 2, 3, 5, 8]);
});

it("allows adding a field with arguments named using a custom inflector", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeAddInflectorsPlugin({
      echoFieldName() {
        return this.camelCase("my-custom-echo-field-name");
      },
    }),
    makeExtendSchemaPlugin((build) => ({
      typeDefs: gql`
        extend type Query {
          """
          Gives you back what you put in
          """
          ${build.inflection.echoFieldName()}(input: [Int!]!): [Int!]!
        }
      `,
      resolvers: {
        Query: {
          [build.inflection.echoFieldName()]: {
            resolve: resolvers.Query.echo,
          },
        },
      },
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      {
        echo: myCustomEchoFieldName(input: [1, 1, 2, 3, 5, 8])
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data.echo).toEqual([1, 1, 2, 3, 5, 8]);
});

it("supports @scope directive with simple values", async () => {
  let scope;
  function storeScope(_scope) {
    if (scope) {
      throw new Error("Scope already stored!");
    }
    scope = _scope;
  }
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          Gives you back what you put in
          """
          echo(input: [Int!]!): [Int!]!
            @scope(
              isEchoField: true
              stringTest: "THIS_IS_A_STRING"
              intTest: 42
              floatTest: 3.141592
              nullTest: null
            )
        }
      `,
      resolvers,
    })),
    TestUtils_ExtractScopePlugin(
      "GraphQLObjectType:fields:field",
      "Query",
      "echo",
      storeScope,
    ),
  ]);
  expect(scope).toBeTruthy();
  expect(scope.isEchoField).toEqual(true);
  expect(scope.stringTest).toEqual("THIS_IS_A_STRING");
  expect(scope.intTest).toEqual(42);
  expect(scope.floatTest).toEqual(3.141592);
  expect(scope.nullTest).toEqual(null);
  expect(scope).toMatchSnapshot();
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      {
        echo(input: [1, 1, 2, 3, 5, 8])
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data.echo).toEqual([1, 1, 2, 3, 5, 8]);
});

it("supports @scope directive with variable value", async () => {
  let scope;
  function storeScope(_scope) {
    if (scope) {
      throw new Error("Scope already stored!");
    }
    scope = _scope;
  }
  const secret = Symbol("test-secret");
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Query {
          """
          Gives you back what you put in
          """
          echo(input: [Int!]!): [Int!]!
            @scope(
              isEchoField: true
              stringTest: "THIS_IS_A_STRING"
              intTest: 42
              floatTest: 3.141592
              nullTest: null
              embedTest: ${embed({
                [secret]: "Fred",
                sub: [
                  [11, 22],
                  [33, 44],
                ],
              })}
            )
        }
      `,
      resolvers,
    })),
    TestUtils_ExtractScopePlugin(
      "GraphQLObjectType:fields:field",
      "Query",
      "echo",
      storeScope,
    ),
  ]);
  expect(scope).toBeTruthy();
  expect(scope.isEchoField).toEqual(true);
  expect(scope.stringTest).toEqual("THIS_IS_A_STRING");
  expect(scope.intTest).toEqual(42);
  expect(scope.floatTest).toEqual(3.141592);
  expect(scope.nullTest).toEqual(null);
  expect(scope.embedTest).toBeTruthy();
  expect(scope.embedTest[secret]).toEqual("Fred");
  expect(scope.embedTest.sub[1][1]).toEqual(44);
  expect(scope).toMatchSnapshot();
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      {
        echo(input: [1, 1, 2, 3, 5, 8])
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data.echo).toEqual([1, 1, 2, 3, 5, 8]);
});

it("supports defining new types", async () => {
  const inputsSeen = [];
  const enumsSeen = [];
  const EchoCount = gql`
    enum EchoCount {
      ONCE
      TWICE
      FOREVER
    }
  `;

  const EchoInput = gql`
    input EchoInput {
      text: String!
      int: Int
      float: Float!
      count: EchoCount = FOREVER
    }
  `;
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        ${EchoCount}
        ${EchoInput}

        type EchoOutput {
          text: String!
          int: Int
          float: Float!
          count: EchoCount!
        }

        extend input EchoInput {
          intList: [Int!]
        }

        extend type EchoOutput {
          intList: [Int!]
        }

        extend type Query {
          """
          Gives you back what you put in
          """
          echo(input: EchoInput, enum: EchoCount = FOREVER): EchoOutput
        }
      `,
      resolvers: {
        EchoCount: {
          FOREVER: "forever and ever and ever",
        },
        Query: {
          echo: {
            resolve(_query, args) {
              inputsSeen.push(args.input);
              enumsSeen.push(args.enum);
              return args.input;
            },
          },
        },
      },
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      {
        t0: echo(input: { text: "Hi0", float: -0.42 }, enum: ONCE) {
          text
          int
          float
          intList
          count
        }
        t1: echo(input: { text: "Hi1", float: 0.23, count: ONCE }) {
          text
          int
          float
          intList
          count
        }
        t2: echo(input: { text: "Hi2", int: 42, float: 1.23, count: TWICE }) {
          text
          int
          float
          intList
          count
        }
        t3: echo(
          input: {
            text: "Hi3"
            int: 88
            float: 2.23
            intList: [99, 22, 33]
            count: FOREVER
          }
        ) {
          text
          int
          float
          intList
          count
        }
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data).toMatchSnapshot();
  expect(inputsSeen.length).toEqual(4);
  expect(enumsSeen.length).toEqual(4);
  expect(inputsSeen.map((s) => s.count)).toEqual([
    "forever and ever and ever",
    "ONCE",
    "TWICE",
    "forever and ever and ever",
  ]);
  expect(enumsSeen).toEqual([
    "ONCE",
    "forever and ever and ever",
    "forever and ever and ever",
    "forever and ever and ever",
  ]);
});

it("supports defining a simple mutation", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Mutation {
          add(a: Int, b: Int): Int
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      mutation {
        add(a: 101, b: 42)
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data).toMatchSnapshot();
});

it("supports defining a more complex mutation", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        input EchoInput {
          text: String!
          int: Int
          float: Float!
          intList: [Int!]
        }

        type EchoOutput {
          text: String!
          int: Int
          float: Float!
          intList: [Int!]
        }

        extend type Mutation {
          """
          Gives you back what you put in
          """
          echo(input: EchoInput): EchoOutput
        }
      `,
      resolvers: {
        Mutation: {
          echo: {
            resolve(_query, args) {
              return args.input;
            },
          },
        },
      },
    })),
  ]);
  expect(schema).toMatchSnapshot();
  const { data, errors } = await graphql(
    schema,
    `
      mutation {
        t1: echo(input: { text: "Hi1", float: 0.23 }) {
          text
          int
          float
          intList
        }
        t2: echo(input: { text: "Hi2", int: 42, float: 1.23 }) {
          text
          int
          float
          intList
        }
        t3: echo(
          input: { text: "Hi3", int: 88, float: 2.23, intList: [99, 22, 33] }
        ) {
          text
          int
          float
          intList
        }
      }
    `,
  );
  expect(errors).toBeFalsy();
  expect(data).toMatchSnapshot();
});

it("supports defining a simple subscription", async () => {
  const schema = await buildSchema([
    ...simplePlugins,
    makeExtendSchemaPlugin((_build) => ({
      typeDefs: gql`
        extend type Subscription {
          clockTicks(
            """
            How frequently to fire a clock tick (milliseconds)
            """
            frequency: Int = 100
          ): Float
        }
      `,
      resolvers,
    })),
  ]);
  expect(schema).toMatchSnapshot();

  // Let's do a standard resolve:
  let before = Date.now();
  expect(timerRunning).toBeFalsy();
  const { data, errors } = await graphql(
    schema,
    `
      subscription {
        clockTicks
      }
    `,
  );
  let after = Date.now();
  expect(after).toBeGreaterThanOrEqual(before + 100);
  expect(errors).toBeFalsy();
  expect(data.clockTicks).toBeGreaterThanOrEqual(before);
  expect(data.clockTicks).toBeLessThanOrEqual(after);
  expect(timerRunning).toBeFalsy();

  // Now lets try subscribing:
  before = Date.now();
  const iterator = await subscribe(
    schema,
    parse(`
      subscription {
        clockTicks(frequency: 50)
      }
    `),
  );
  after = Date.now();

  // expect(iterator).toBeInstanceOf(AsyncIterator);
  expect(iterator.errors).toBeFalsy();
  expect(timerRunning).toBeTruthy();

  // Lets get the next 5 values
  let lastTick = before;
  for (let i = 0; i < 5; i++) {
    const { value, done } = await iterator.next();
    expect(done).toBeFalsy();
    const { data, errors } = value;
    expect(errors).toBeFalsy();
    const currentTick = data.clockTicks;
    expect(currentTick).toBeTruthy();
    expect(currentTick).toBeGreaterThanOrEqual(lastTick + 49);
    lastTick = currentTick;
  }
  expect(timerRunning).toBeTruthy();

  // And now stop the iterator
  await iterator.return();
  expect(timerRunning).toBeFalsy();
  const { value, done } = await iterator.next();
  expect(done).toBeTruthy();
  expect(value).toBe(undefined);
});
