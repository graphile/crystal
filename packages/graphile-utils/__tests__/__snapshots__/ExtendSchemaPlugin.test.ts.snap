// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts an array of typedefs 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """A random number generated by a fair dice roll."""
  randomNumber: Int!

  """Gives a list of numbers that were randomly generated by fair dice roll"""
  randomNumbers: [Int!]!
}

`;

exports[`allows adding a field with arguments 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives you back what you put in"""
  echo(input: [Int!]!): [Int!]!
}

`;

exports[`allows adding a field with arguments named using a custom inflector 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives you back what you put in"""
  myCustomEchoFieldName(input: [Int!]!): [Int!]!
}

`;

exports[`allows adding a non-null list of non-null type 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives a list of numbers that were randomly generated by fair dice roll"""
  randomNumbers: [Int!]!
}

`;

exports[`allows adding a non-null type 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """A random number generated by a fair dice roll."""
  randomNumber: Int!
}

`;

exports[`allows adding a simple type 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """A random number generated by a fair dice roll."""
  randomNumber: Int
}

`;

exports[`supports @scope directive with simple values 1`] = `
Object {
  "__origin": "graphile-build built-in (root query type)",
  "fieldDirectives": Object {
    "scope": Object {
      "floatTest": 3.141592,
      "intTest": 42,
      "isEchoField": true,
      "nullTest": null,
      "stringTest": "THIS_IS_A_STRING",
    },
  },
  "fieldName": "echo",
  "floatTest": 3.141592,
  "intTest": 42,
  "isEchoField": true,
  "isRootQuery": true,
  "nullTest": null,
  "stringTest": "THIS_IS_A_STRING",
  Symbol(hints): Object {
    "fieldDirectives": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "fieldName": "Within context for GraphQLObjectType 'Query'",
    "floatTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "intTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "isEchoField": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "nullTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "stringTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
  },
}
`;

exports[`supports @scope directive with simple values 2`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives you back what you put in"""
  echo(input: [Int!]!): [Int!]!
}

`;

exports[`supports @scope directive with variable value 1`] = `
Object {
  "__origin": "graphile-build built-in (root query type)",
  "embedTest": Object {
    "sub": Array [
      Array [
        11,
        22,
      ],
      Array [
        33,
        44,
      ],
    ],
    Symbol(test-secret): "Fred",
  },
  "fieldDirectives": Object {
    "scope": Object {
      "embedTest": Object {
        "sub": Array [
          Array [
            11,
            22,
          ],
          Array [
            33,
            44,
          ],
        ],
        Symbol(test-secret): "Fred",
      },
      "floatTest": 3.141592,
      "intTest": 42,
      "isEchoField": true,
      "nullTest": null,
      "stringTest": "THIS_IS_A_STRING",
    },
  },
  "fieldName": "echo",
  "floatTest": 3.141592,
  "intTest": 42,
  "isEchoField": true,
  "isRootQuery": true,
  "nullTest": null,
  "stringTest": "THIS_IS_A_STRING",
  Symbol(hints): Object {
    "embedTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "fieldDirectives": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "fieldName": "Within context for GraphQLObjectType 'Query'",
    "floatTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "intTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "isEchoField": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "nullTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
    "stringTest": "Extending scope for field 'echo' within context for GraphQLObjectType 'Query'",
  },
}
`;

exports[`supports @scope directive with variable value 2`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives you back what you put in"""
  echo(input: [Int!]!): [Int!]!
}

`;

exports[`supports defining a more complex mutation 1`] = `
input EchoInput {
  text: String!
  int: Int
  float: Float!
  intList: [Int!]
}

type EchoOutput {
  text: String!
  int: Int
  float: Float!
  intList: [Int!]
}

"""
The root mutation type which contains root level fields which mutate data.
"""
type Mutation {
  """Gives you back what you put in"""
  echo(input: EchoInput): EchoOutput
}

"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!
}

`;

exports[`supports defining a more complex mutation 2`] = `
Object {
  "t1": Object {
    "float": 0.23,
    "int": null,
    "intList": null,
    "text": "Hi1",
  },
  "t2": Object {
    "float": 1.23,
    "int": 42,
    "intList": null,
    "text": "Hi2",
  },
  "t3": Object {
    "float": 2.23,
    "int": 88,
    "intList": Array [
      99,
      22,
      33,
    ],
    "text": "Hi3",
  },
}
`;

exports[`supports defining a simple mutation 1`] = `
"""
The root mutation type which contains root level fields which mutate data.
"""
type Mutation {
  add(a: Int, b: Int): Int
}

"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!
}

`;

exports[`supports defining a simple mutation 2`] = `
Object {
  "add": 143,
}
`;

exports[`supports defining a simple subscription 1`] = `
"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!
}

"""
The root subscription type: contains realtime events you can subscribe to with the \`subscription\` operation.
"""
type Subscription {
  clockTicks(
    """How frequently to fire a clock tick (milliseconds)"""
    frequency: Int = 100
  ): Float
}

`;

exports[`supports defining new types 1`] = `
enum EchoCount {
  ONCE
  TWICE
  FOREVER
}

input EchoInput {
  text: String!
  int: Int
  float: Float!
  count: EchoCount = FOREVER
  intList: [Int!]
}

type EchoOutput {
  text: String!
  int: Int
  float: Float!
  count: EchoCount!
  intList: [Int!]
}

"""The root query type which gives access points into the data universe."""
type Query {
  """
  Exposes the root query type nested one level down. This is helpful for Relay 1
  which can only query top level fields if they are in a particular form.
  """
  query: Query!

  """Gives you back what you put in"""
  echo(input: EchoInput, enum: EchoCount = FOREVER): EchoOutput
}

`;

exports[`supports defining new types 2`] = `
Object {
  "t0": Object {
    "count": "FOREVER",
    "float": -0.42,
    "int": null,
    "intList": null,
    "text": "Hi0",
  },
  "t1": Object {
    "count": "ONCE",
    "float": 0.23,
    "int": null,
    "intList": null,
    "text": "Hi1",
  },
  "t2": Object {
    "count": "TWICE",
    "float": 1.23,
    "int": 42,
    "intList": null,
    "text": "Hi2",
  },
  "t3": Object {
    "count": "FOREVER",
    "float": 2.23,
    "int": 88,
    "intList": Array [
      99,
      22,
      33,
    ],
    "text": "Hi3",
  },
}
`;
